elastio-s3-changelog/cloudformation-single.yaml

```
@@ -0,0 +1,263 @@
+    Description: >
+      If set to true, enables quasi-realtime scanning.
+      If set to false, disables quasi-realtime scanning.
+
+  KeyPrefixes:
+    # Unfortunately, it's impossible to pass CommaDelimitedList to a nested stack,
+    # so we have to use String here, and then split it manually.
+    Type: String
+    AllowedPattern: ^(([^/,\s]+/)+)(,([^/,\s]+/)+)*$|^$
+    ConstraintDescription: Prefixes must end (but not start) with '/', and cannot contain spaces.
+    Description: >
+      Comma-delimited list of prefixes of objects to scan. If empty, all objects will be scanned.
+      Example: xyz/,foo/bar/
+
+Conditions:
+  ScanExistingObjects: !Equals [!Ref ScanExistingObjects, 'true']
+  EnableQrts: !Equals [!Ref EnableQrts, 'true']
+  NoKeyPrefixes: !Equals [!Ref KeyPrefixes, '']
+
+Resources:
+  queue:
+    Type: AWS::SQS::Queue
+    Properties:
+      MessageRetentionPeriod: 1209600 # 14 days
+      VisibilityTimeout: 43200 # 12 hours
+      RedrivePolicy:
+        deadLetterTargetArn: !GetAtt dlq.Arn
+        maxReceiveCount: 3
+      Tags:
+        - Key: elastio:resource
+          Value: 'true'
+        - Key: elastio:s3-changelog-bucket
+          Value: !Ref BucketName
+        - Key: elastio:s3-changelog-stack-arn
+          Value: !Ref AWS::StackId
+
+  dlq:
+    Type: AWS::SQS::Queue
+    Properties:
+      MessageRetentionPeriod: 1209600
+      VisibilityTimeout: 43200
+      Tags:
+        - Key: elastio:resource
+          Value: 'true'
+        - Key: elastio:s3-changelog-dlq
+          Value: !Ref BucketName
+        - Key: elastio:s3-changelog-stack-arn
+          Value: !Ref AWS::StackId
+
+  eventRule:
+    Type: AWS::Events::Rule
+    DependsOn: queue
+    Properties:
+      Description: !Sub Track new objects creation in ${BucketName}
+      EventPattern: !If
+        - NoKeyPrefixes
+        - !Sub >
+          {
+            "source": ["aws.s3"],
+            "detail-type": ["Object Created"],
+            "detail": { "bucket": { "name": ["${BucketName}"] } }
+          }
+        - !Join
+          - ''
+          - - !Sub >
+              {
+                "source": ["aws.s3"],
+                "detail-type": ["Object Created"],
+                "detail":
+                {
+                  "bucket": { "name": ["${BucketName}"] },
+                  "object": { "key": [
+            - !Sub
+              - '{ "prefix": "${prefixes}" }'
+              - prefixes: !Join
+                - '" },{ "prefix": "'
+                - !Split [',', !Ref KeyPrefixes]
+            - '] } } }'
+      Targets:
+        - Arn: !GetAtt queue.Arn
+          Id: changelog-queue
+          InputTransformer:
+            InputPathsMap:
+              key: '$.detail.object.key'
+              type: '$.detail-type'
+            InputTemplate:
+              '{
+                "objects": [{
+                  "key": <key>,
+                  "event_type": <type>
+                }]
+              }'
+
+  queueEventBridgePolicy:
```

Keep the name more generic. What if we extend this inline policy in the future?
```suggestion
  queueInlinePolicy:
```